File: src/lib/tauri/ssh.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type {
  SshConnectInput,
  SshInputPacket,
  SshLifecycleEvent,
  SshOutputEvent,
  SshResizeInput,
  SshSessionSummary
} from "../../types/ssh";

export const SSH_OUTPUT_EVENT = "ssh-output";
export const SSH_LIFECYCLE_EVENT = "ssh-lifecycle";

export async function connectSshSession(
  input: SshConnectInput
): Promise<SshSessionSummary> {
  return invoke<SshSessionSummary>("connect_ssh_session", { input });
}

export async function sendSshInput(packet: SshInputPacket): Promise<void> {
  return invoke("send_ssh_input", { packet });
}

export async function resizeSshSession(input: SshResizeInput): Promise<void> {
  return invoke("resize_ssh_session", { input });
}

export async function disconnectSshSession(sessionId: string): Promise<void> {
  return invoke("disconnect_ssh_session", { sessionId });
}

export async function listSshSessions(): Promise<SshSessionSummary[]> {
  return invoke<SshSessionSummary[]>("list_ssh_sessions");
}

export async function listenSshOutput(
  handler: (event: SshOutputEvent) => void
): Promise<UnlistenFn> {
  return listen<SshOutputEvent>(SSH_OUTPUT_EVENT, ({ payload }) => handler(payload));
}

export async function listenSshLifecycle(
  handler: (event: SshLifecycleEvent) => void
): Promise<UnlistenFn> {
  return listen<SshLifecycleEvent>(SSH_LIFECYCLE_EVENT, ({ payload }) =>
    handler(payload)
  );
}

File: src-tauri/src/commands/ssh.rs
use crate::{
    app_state::AppState,
    modules::models::{SshConnectInput, SshInputPacket, SshResizeInput, SshSessionSummary},
};
use tauri::State;

#[tauri::command]
pub async fn connect_ssh_session(
    input: SshConnectInput,
    state: State<'_, AppState>,
) -> Result<SshSessionSummary, String> {
    state
        .ssh
        .connect(input)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn send_ssh_input(
    packet: SshInputPacket,
    state: State<'_, AppState>,
) -> Result<(), String> {
    state
        .ssh
        .send_input(&packet.session_id, &packet.data)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn resize_ssh_session(
    input: SshResizeInput,
    state: State<'_, AppState>,
) -> Result<(), String> {
    state
        .ssh
        .resize(
            &input.session_id,
            input.cols,
            input.rows,
            input.pixel_width.unwrap_or_default(),
            input.pixel_height.unwrap_or_default(),
        )
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn disconnect_ssh_session(
    session_id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    state
        .ssh
        .disconnect(&session_id)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn list_ssh_sessions(
    state: State<'_, AppState>,
) -> Result<Vec<SshSessionSummary>, String> {
    state
        .ssh
        .list_sessions()
        .await
        .map_err(|error| error.to_string())
}
